### PR TITLE
Fix scalars demo summary writing race condition

### PR DESCRIPTION
**Pull request by nfelt**
This makes the scalars demo generate 100% deterministic deterministic summary output, by correcting for a race condition that probably afflicts a significant amount of TensorFlow summary-writing code.

The general race condition occurs when summary ops are fetched (in a `sess.run()` call) alongside ops that mutate a variable on which the actual summary values depend.  The scalars demo is a particularly simple example of this (there's a temperature variable that is incremented in each step, and also written out as a scalar summary), but many kinds of other TF graphs have a fundamentally similar structure. [1]

The problem lies in the fact that a `sess.run()` call with multiple fetches doesn't guarantee to execute the fetches in any particular order beyond what are expressed in explicit dependencies (either data dependencies, or "manual" control dependencies).  And in particular, if a leaf node (like typical summary ops) depends directly on a variable, it can be executed before *or* after the variable is updated - and indeed in practice the ordering appears to be non-deterministic. [2]

The straightforward fix (and the one I use here) is adding a control dependency so that the variable update op depends on the summary ops that use the original variable value, ensuring that the summaries are generated using the pre-update value.  If instead one wants summaries to report the post-update value, control dependencies are still an option, but it's also easy to achieve that by just dangling the summary ops off the new value, after decomposing the update like this:

```python
new_value = compute_new_value(var)
tb.summary.scalar(new_value)
var.assign(new_value)
```

I believe a similar data-dependency-only strategy would also be possible for summaries using the pre-update value by just inserting a `tf.identity` op to "cache" the original variable value V in a new tensor V', then dangling the summaries and the computation of the new value off of V' - the identity op result will be computed before the variable update due to the data dependency, and since multiple fetches in `sess.run()` do guarantee to execute any shared op in the graph exactly once, the summary op will also pick up V' with the original variable value - though I haven't tested this approach, and it's a bit more subtle/tricky than the control dependency approach.

[1] I tried to reproduce this race condition with the mnist tutorial code and failed to do so, but I suspect it's still affected in theory - I think it only wasn't showing the race because the variable update for even a relatively simple graph like the mnist tutorial one is expensive enough that it always seems to execute after the summary ops.

[2] See for more details:
https://stackoverflow.com/questions/37854477/fixing-race-condition-in-tensorflow-run
https://www.quora.com/Does-TensorFlow-evaluate-common-nodes-to-multiple-ops-multiple-times

### CODE REVIEW

File: tensorboard/plugins/scalar/scalars_demo.py [hunk 1]
```diff
@@ -92,12 +92,14 @@ def run(logdir, run_name,
...
                          'step, in Kelvins.')
 
-  # Now, augment the current temperature by this delta that we computed.
-  update_step = temperature.assign_add(delta)
-
   # Collect all the scalars that we want to keep track of.
   summ = tf.summary.merge_all()
 
+  # Now, augment the current temperature by this delta that we computed,
+  # blocking the assignment on summary collection to avoid race conditions.
```
Comment (chihuahua) [inline tensorboard/plugins/scalar/scalars_demo.py]: Just to hammer the point, maybe clarify here that we use summary values computed before the update occurs.

**Comment by nfelt** [comment]
Yeah, I think the `tf.identity` pattern would get a little weird, although the control dependencies are also a little ugly.

I'm not sure why TF variables weren't designed with some better way to explicitly express a dependency on their original values (as of the start of `sess.run()`).  In fact there's no real way to do that as far as I can tell short of the `tf.identity` hack, since for the control dependency version, you need to know all the ops that might update the variables, and if you add a new op and forget to include it you're back to the race condition.  You'd think that this would be a more common concern, but I guess it just doesn't come up frequently enough.

**Comment by chihuahua** [comment]
Indeed ...

Some abstractions atop TensorFlow do incorporate some notion of separating updating values vs reading them. For instance, streaming metrics return an op that contains results as well as an update op. A user can then decide how often and when to update the value vs reading the value:

```python
unused_summary_op, update_op = summary_lib.pr_curve_streaming_op('pr_curve', ...)
...
writer = tf.summary.FileWriter(log_dir, graph)
for i in xrange(NUM_STEPS):
  sess.run(update_op)
  if i % 5 == 0:
    # Only write summaries every 5 steps.
    writer.add_summary(sess.run(merged_summary_op), i)
```

**Comment by nfelt** [comment]
Yeah, that's true - that is the other option, is to explicitly run the summary ops in one `sess.run` call and then the update ops in a separate `sess.run` call.  In a way, that's the most obvious solution, and gives you some extra flexibility (e.g. you can record summaries at a step 0 before any updates, and then record them after the update for steps 1...N).  But I didn't want to recommend that in general because separate `sess.run()` calls might mean re-executed shared nodes unintentionally, which could be an extra performance cost or in the worst cases lead to multiple updates (if any of your summaries ops depend on stateful updates that are also depended on by the update op).

**Comment by chihuahua** [comment]
That makes sense to me. If dependencies in the TensorFlow graph are awry, the ops could lead to duplicate computations - most test suites for streaming metrics have an `isIdempotent` test.

And agreed, using update ops seems like overkill to recommend for this demo :)

**Pull request merged**